sql/core/src/main/scala/org/apache/spark/sql/catalyst/analysis/ResolveSessionCatalog.scala

```diff
@@ -657,14 +659,8 @@ class ResolveSessionCatalog(
       comment: Option[String],
       storageFormat: CatalogStorageFormat,
       external: Boolean): CatalogTable = {
-    if (external) {
-      if (DDLUtils.isHiveTable(Some(provider))) {
-        if (location.isEmpty) {
-          throw new AnalysisException(s"CREATE EXTERNAL TABLE must be accompanied by LOCATION")
-        }
-      } else {
-        throw new AnalysisException(s"Operation not allowed: CREATE EXTERNAL TABLE ... USING")
-      }
+    if (external && location.isEmpty) {
+      throw new AnalysisException(s"CREATE EXTERNAL TABLE must be accompanied by LOCATION")
```

Comment:
The PR description says "This PR allows CREATE EXTERNAL TABLE when LOCATION is present" but the effect of this change is to disallow `EXTERNAL` when `LOCATION` is not present.

Since the session catalog may be wrapped by a v2 catalog, this disallows `EXTERNAL` for some v2 catalog plugin use cases. That violates the v2 behavior expectations because `EXTERNAL` should be passed to a v2 catalog unless it would have failed before. I don't think this change is correct.

Also, why does this change need to be included to change the default provider? If it doesn't need to be included, then it should be in a separate PR.

Reply:
> but the effect of this change is to disallow EXTERNAL when LOCATION is not present.

This is already the case before this PR. The error message is pretty clear about it: `CREATE EXTERNAL TABLE must be accompanied by LOCATION`. The change here only loose the restrictions.

I include this change because otherwise tests that use `CREATE EXTERNAL TABLE ... LOCATION` without USING or STORED AS are failing.

Reply:
Before, this was enforcing that `EXTERNAL` could not be combined with `USING`, which was the previous behavior.

Now, it is enforcing that `EXTERNAL` cannot be present without `LOCATION` for data source tables, which has been discussed by the community and decided against. I think this should be left alone, or that it should be enforced in the session catalog rather than the analyzer rules.

Reply:
Please correct me if I read the code wrongly. Previously Spark fails when EXTERNAL is specified and the following two conditions are met:
1. it's a hive table and LOCATION is not specified.
2. it's a data source table.

That said, we had a big restriction before that EXTERNAL is forbidden completely for data source tables, no matter LOCATION is present or not. Now my change allows it if LOCATION is present.

Reply:
I get your point that we want to allow the session catalog plugin to overwrite this restriction. I think it's as easy as moving the check to `V2SessionCatalog`, but then it's not very related to this PR.

Reply:
We might as well delegate this to the catalog if we're going to bother changing this code path as part of this PR.

Reply:
This PR adds support for `EXTERNAL` with data source tables, but only for tables with the `LOCATION` clause. I have no problem with adding that support, or for enforcing the `LOCATION` restriction in Spark's catalog.

The problem is that it is done in the analyzer rather than in the catalog implementation, so the `LOCATION` clause restriction is applied before calling a v2 extension to the session catalog.

We've now had several problems with behavior mismatches between the session catalog and other v2 catalogs, so I think that support for `EXTERNAL` should be added properly: pass the flag correctly to the v2 extension and enforce the `LOCATION` restriction in the catalog rather than in the analysis rules. I'd also like to have tests for the newly supported SQL cases to validate that behavior.

I know that the reason behind this was to avoid test failures, but I think only partially adding this feature means that it isn't done right: it introduces a behavior difference between session and other v2 catalogs, and would be missing tests.

If you want to update this to add the `LOCATION` restriction in the session catalog, pass the v2 property correctly to catalog extensions, and add tests, then I'm okay with this change. However, then we would be looking at a significantly larger PR, so it is unlikely that this should be done at the same time as adding configuration to use data source tables by default.

Reply:
> but I think only partially adding this feature means that it isn't done right

OK now we are on the same page. This change doesn't add new restriction, but you think we should loose the restriction in a more proper way that works better with session catalog plugin.

I agree with it, and I'm going to revert this change (since I fliped the config) and do it later.

Reply:
Thanks for reverting this. Let me know and I can help review the addition later.